src/Processors/Transforms/MergeJoinTransform.cpp

```diff
@@ -0,0 +1,865 @@
+        prepareChunk(inputs[i].chunk);
+        copyColumnsResized(inputs[i].chunk.getColumns(), 0, 0, sample_chunks.emplace_back());
+        consume(inputs[i], i);
+    }
+}
+
+void MergeJoinAlgorithm::consume(Input & input, size_t source_num)
+{
+    if (input.skip_last_row)
+        throw Exception("skip_last_row is not supported", ErrorCodes::NOT_IMPLEMENTED);
+
+    if (input.permutation)
+        throw DB::Exception("permutation is not supported", ErrorCodes::NOT_IMPLEMENTED);
+
+    if (input.chunk)
+    {
+        stat.num_blocks[source_num] += 1;
+        stat.num_rows[source_num] += input.chunk.getNumRows();
+    }
+
+    prepareChunk(input.chunk);
+    cursors[source_num]->setChunk(std::move(input.chunk));
+}
+
+template <JoinKind kind>
+struct AllJoinImpl
+{
+    constexpr static bool enabled = isInner(kind) || isLeft(kind) || isRight(kind) || isFull(kind);
+
+    static void join(FullMergeJoinCursor & left_cursor,
+                     FullMergeJoinCursor & right_cursor,
+                     size_t max_block_size,
+                     PaddedPODArray<UInt64> & left_map,
+                     PaddedPODArray<UInt64> & right_map,
+                     std::unique_ptr<AllJoinState> & state)
+    {
+        right_map.clear();
+        right_map.reserve(max_block_size);
+        left_map.clear();
+        left_map.reserve(max_block_size);
+
+        size_t rpos = std::numeric_limits<size_t>::max();
+        size_t lpos = std::numeric_limits<size_t>::max();
+        int cmp = 0;
+        assert(left_cursor->isValid() && right_cursor->isValid());
+        while (left_cursor->isValid() && right_cursor->isValid())
+        {
+            lpos = left_cursor->getRow();
+            rpos = right_cursor->getRow();
+
+            cmp = compareCursors(left_cursor.cursor, right_cursor.cursor);
+            if (cmp == 0)
+            {
+                size_t lnum = nextDistinct(left_cursor.cursor);
+                size_t rnum = nextDistinct(right_cursor.cursor);
+
+                bool all_fit_in_block = std::max(left_map.size(), right_map.size()) + lnum * rnum <= max_block_size;
+                bool have_all_ranges = left_cursor.cursor.isValid() && right_cursor.cursor.isValid();
+                if (all_fit_in_block && have_all_ranges)
+                {
+                    /// fast path if all joined rows fit in one block
+                    for (size_t i = 0; i < rnum; ++i)
+                    {
+                        addRange(left_map, lpos, left_cursor.cursor.getRow());
+                        addMany(right_map, rpos + i, lnum);
+                    }
+                }
+                else
+                {
+                    assert(state == nullptr);
+                    state = std::make_unique<AllJoinState>(left_cursor.cursor, lpos, right_cursor.cursor, rpos);
+                    state->addRange(0, left_cursor.getCurrent().clone(), lpos, lnum);
```

Comment:
Shall we only copy the necessary rows from left_cursor.getCurrent() chunk and right_cursor.getCurrent() chunk?